### PR TITLE
wip: Converts Todo sample from onEvent to makeSink.

### PR DESCRIPTION
**Pull request by rjrjr**

### CODE REVIEW

File: kotlin/samples/todo-android/common/src/main/java/com/squareup/sample/todo/TodoEditorWorkflow.kt [hunk 1]
```diff
@@ -36,50 +38,84 @@ data class TodoRow(
...
-              copy(done = !done)
-            }))
-            is TextChanged -> emitOutput(ListUpdated(input.updateRow(it.index) {
-              copy(text = it.text)
-            }))
-            is DeleteClicked -> emitOutput(ListUpdated(input.removeRow(it.index)))
-          }
-        }
+        input.copy(rows = input.rows + TodoRow("")),
+        { sink.send(TitleChanged(input, it)) },
```
Comment (zach-klippenstein) [inline kotlin/samples/todo-android/common/src/main/java/com/squareup/sample/todo/TodoEditorWorkflow.kt]: These should use named params.

File: kotlin/samples/todo-android/common/src/main/java/com/squareup/sample/todo/TodoEditorWorkflow.kt [hunk 2]
```diff
@@ -36,50 +38,84 @@ data class TodoRow(
...
+
+    // Make event handling idempotent.
+    var eventFired = false
+    val sink = object : Sink<TodoAction> {
+      override fun send(value: TodoAction) {
+        if (eventFired) return
+        eventFired = true
+        rawSink.send(value)
+      }
+    }
```
Comment (zach-klippenstein) [inline kotlin/samples/todo-android/common/src/main/java/com/squareup/sample/todo/TodoEditorWorkflow.kt]: File an issue with a TODO here to implement the same "if the sink type is the same across render passes, let multiple calls to the same sink instance work" behavior that Swift has.

File: kotlin/samples/todo-android/common/src/main/java/com/squareup/sample/todo/TodoListsWorkflow.kt [hunk 3]
```diff
@@ -29,8 +30,8 @@ class TodoListsWorkflow : StatelessWorkflow<List<TodoList>, Int, TodoListsRender
     input: List<TodoList>,
     context: RenderContext<Nothing, Int>
   ): TodoListsRendering {
-    return TodoListsRendering(input, context.onEvent { row ->
-      emitOutput(row)
-    })
+    val sink = context.makeSink<WorkflowAction<Nothing, Int>>()
+
+    return TodoListsRendering(input) { sink.send(emitOutput(it)) }
```
Comment (zach-klippenstein) [inline kotlin/samples/todo-android/common/src/main/java/com/squareup/sample/todo/TodoListsWorkflow.kt]: Discussed offline: `makeSink` variant with anonymous action lambda.

File: kotlin/samples/todo-android/common/src/main/java/com/squareup/sample/todo/TodoEditorWorkflow.kt (hunk 2 above)
Comment (rjrjr) [reply kotlin/samples/todo-android/common/src/main/java/com/squareup/sample/todo/TodoEditorWorkflow.kt]: https://github.com/square/workflow/issues/541

File: kotlin/samples/todo-android/common/src/main/java/com/squareup/sample/todo/TodoListsWorkflow.kt (hunk 3 above)
Comment (rjrjr) [reply kotlin/samples/todo-android/common/src/main/java/com/squareup/sample/todo/TodoListsWorkflow.kt]:
```
    // A sink that emits the given index as the result of this workflow.
      emitOutput(row)	    val sink: Sink<Int> = context.makeEventSink { index: Int ->
    })	      index
    }
    return TodoListsRendering(lists = input, onRowClicked = sink::send)
```

**Pull request merged**